<div class='alert alert-warning'>

SciPy's interactive examples with Jupyterlite are experimental and may not always work as expected. Execution of cells containing imports may result in large downloads (up to 60MB of content for the first import from SciPy). Load times when importing from SciPy may take roughly 10-20 seconds. If you notice any problems, feel free to open an [issue](https://github.com/scipy/scipy/issues/new/choose).

</div>

In [ ]:
import numpy as np
from scipy.linalg import schur, eigvals
A = np.array([[0, 2, 2], [0, 1, 2], [1, 0, 1]])
T, Z = schur(A)
T

array([[ 2.65896708,  1.42440458, -1.92933439],
       [ 0.        , -0.32948354, -0.49063704],
       [ 0.        ,  1.31178921, -0.32948354]])

In [ ]:
Z

array([[0.72711591, -0.60156188, 0.33079564],
       [0.52839428, 0.79801892, 0.28976765],
       [0.43829436, 0.03590414, -0.89811411]])

In [ ]:
T2, Z2 = schur(A, output='complex')
T2

array([[ 2.65896708, -1.22839825+1.32378589j,  0.42590089+1.51937378j], # may vary
       [ 0.        , -0.32948354+0.80225456j, -0.59877807+0.56192146j],
       [ 0.        ,  0.                    , -0.32948354-0.80225456j]])

In [ ]:
eigvals(T2)

array([2.65896708, -0.32948354+0.80225456j, -0.32948354-0.80225456j])   # may vary

A custom eigenvalue-sorting condition that sorts by positive imaginary part
is satisfied by only one eigenvalue.


In [ ]:
_, _, sdim = schur(A, output='complex', sort=lambda x: x.imag > 1e-15)
sdim

1

When ``output='real'`` and the array `a` is real, the `sort` callable must accept
the real and imaginary parts as separate arguments. Note that now the complex
eigenvalues ``-0.32948354+0.80225456j`` and ``-0.32948354-0.80225456j`` will be
treated as a complex conjugate pair, and according to the `sdim` documentation,
complex conjugate pairs for which the condition is True for *either* eigenvalue
increase `sdim` by *two*.


In [ ]:
_, _, sdim = schur(A, output='real', sort=lambda x, y: y > 1e-15)
sdim

2